File: sol/solution/0001-0100/0039. Combination Sum/Solution.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        result = []
        current = []
        def backtrack(start: int, remaining: int):
            if remaining == 0:
                result.append(list(current))
                return
            for i in range(start, len(candidates)):
                num = candidates[i]
                if num > remaining:
                    break
                current.append(num)
                backtrack(i, remaining - num)
                current.pop()
        backtrack(0, target)
        return result

File: sol/solution/0001-0100/0039. Combination Sum/Solution.php
class Solution {
    /**
     * @param Integer[] $candidates
     * @param Integer $target
     * @return Integer[][]
     */
    function combinationSum($candidates, $target) {
        sort($candidates);
        $result = [];
        $current = [];
        $backtrack = function($start, $remaining) use (&$backtrack, &$result, &$current, $candidates) {
            if ($remaining == 0) {
                $result[] = $current;    
                return;
            }
            for ($i = $start; $i < count($candidates); $i++) {
                $num = $candidates[$i];
                if ($num > $remaining) {
                    break; 
                }
                $current[] = $num;
                $backtrack($i, $remaining - $num);
                array_pop($current);              
            }
        };
        $backtrack(0, $target);
        return $result;
    }
}